src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1322,6 +1322,37 @@ static ErrorCode getRefMismatchErrorCode(TypeKind type)
                 };
         }
 
+        public static void CheckDefaultParameterMatch(ErrorCode code, ParameterSymbol source, ParameterSymbol target, Location location, BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(code is ErrorCode.ERR_OptionalParamValueMismatch or ErrorCode.WRN_OptionalParamValueMismatch);
+
+            var sourceParamDefault = source.ExplicitDefaultConstantValue;
+            if (sourceParamDefault is not null && !sourceParamDefault.IsBad)
+            {
+                var targetParamDefault = target.ExplicitDefaultConstantValue;
+                if (targetParamDefault?.IsBad != true && sourceParamDefault != targetParamDefault)
+                {
+                    // Parameter {0} has default value '{1}' in (lambda|method group) and '{2}' in target delegate type.
+                    diagnostics.Add(code, location, source.Ordinal + 1, sourceParamDefault, targetParamDefault ?? ((object)MessageID.IDS_Missing.Localize()));
+                }
+            }
+        }
+
+        private static void CheckForDefaultParameterMismatch(SyntaxNode methodGroupSyntax, MethodSymbol sourceMethod, MethodSymbol targetDelegateInvoke, bool isExtensionMethod, BindingDiagnosticBag diagnostics)
+        {
+            var start = isExtensionMethod ? 1 : 0;
+            Debug.Assert(sourceMethod.ParameterCount == targetDelegateInvoke.ParameterCount + start);
+
+            for (int i = start; i < sourceMethod.ParameterCount; i++)
+            {
+                var delegateParameter = targetDelegateInvoke.Parameters[isExtensionMethod ? i - 1 : i];
```

Comment:
>isExtensionMethod [](http://example.com/codeflow?start=72&length=17)

Can we use `targetDelegateInvoke.ParameterCount` for the `for` loop, and remove the extra check of `isExtensionMethod` here?
```csharp
for (int i = 0; i < targetDelegateInvoke.ParameterCount; i++)
{
    var delegateParameter.Parameters[i];
    ...
    CheckDefaultParameterMatch(..., sourceMethod.Parameters[i + start], ...);
}
```